Add "Hold Off" mod (no long notes)

Spooghetti420 opened the pull request:
Hello, I have made an osu!mania mod which converts all held notes into single-tap notes, one at the start and one at the end. It has a customization which allows users to select how long a held note needs to be for an end note to be added, if at all.
I saw this mod concept while my friend was playing Quaver and thought it might be a fun addition to osu! too. I hope the code is of acceptable quality, though I wasn't sure how to write a test, so sorry for any inconveniences.
I also don't feel awfully confident about the icon that represents the mod, nor the descriptions/tooltips for customization, so, by all means, please change them as necessary.
Thanks

Spooghetti420 committed:
Add No Long Notes mod

Spooghetti420 committed:
Merge branch 'master' of https://github.com/ppy/osu into no-long-notes-mod

Spooghetti420 committed:
Fix newline style in mod

Spooghetti420 committed:
Add another newline

frenzibyte commented on osu.Game.Rulesets.Mania/Mods/ManiaModNoLongNotes.cs:
```diff
@@ -0,0 +1,86 @@
+            Value = true
+        };
+
+        [SettingSource("Length threshold", "Only add an end note for held notes longer than this threshold (in milliseconds)")]
+        public BindableNumber<double> Threshold { get; } = new BindableDouble
+        {
+            MinValue = 1.0,
+            MaxValue = 1990.0,
+            Default = 200.0,
+            Value = 200.0,
+            Precision = 1.0,
```
I'm not really sure this can be exposed to the user. Should be a constant value instead.

bdach replied:
agreed, don''t see much of a point having this as a setting either. especially with these min-max bounds (why 1ms-1990ms?)

Spooghetti420 replied:
I did also feel a little conflicted adding it as a setting, since even I couldn't understand how the "threshold" was even supposed to work at first. If we revert it to a constant, would the 200ms be good, or some other value? 
I only added this setting out of concerns of excessive difficulty, since high-speed maps tend to get a lot of note clutter with too many end notes — hence why I thought having an option to control it might be more helpful, since it can be adapted to the difficulty of the map. I couldn't think of a good way to make this depend on how hard the map was *supposed* to be tho, so I just let the user decide. Probably overthinking this at any rate.

smoogipoo replied:
A threshold/ms time value doesn't make sense since it bypasses beat snapping, so I would hope that that's used instead. In other rulesets, "taps" are typically 1/1 or 1/2 beats so I would propose the same value to be used here - that is, a hold note shorter than 1/2 beat is converted into only 1 note.

Perhaps if a setting is needed, that could be the setting. "Beat snap of notes generated from long notes" or something like that.

peppy replied:
I'd probably drop this setting and keep things simple to start with?

smoogipoo replied:
Yeah, again I'd pick 1/1 or 1/2 beats to start off with and remove _both_ settings. 1/4 is doable (in my rusty state I can do ~140BPM 1/4s single tap), but that's already towards the upper limit for most people that I don't think will fit well in the general case.

Spooghetti420 replied:
So should we remove the option to choose entirely? Or restrict it to more reasonable values. Forgive me for forgetting about this post, as I went and added up to 1/16 notes, which is way beyond the "doable" 1/4, so if it's like you say, I can't see this getting used by 99% of people, I suppose... shall we remove the settings menu after all?

bdach replied:
I'd still agree with removing the option, personally. Seems like too much customisation just for the sake of it, and it can be re-added in the future if there is actual demand for it.

peppy replied:
Agree with removing and waiting for there to be a demand for it. Choosing a sane setting sounds best.

My rationale is that adding settings like this come with an overhead of supporting them properly (ie `osu-web`, replays etc) and for a mod like this which is bound to see regular use, keeping the behaviour simple is in our best interest.

You're welcome to leave the functions as they are in case we need to add more flexibility in the future - just remove the enum and bindable and replace with a `const` for now.

Spooghetti420 replied:
Thanks, this seems much more sensible. I got rid of the settings and updated the tests, so now we have it set to half-note as a constant.

frenzibyte commented on osu.Game.Rulesets.Mania/Mods/ManiaModNoLongNotes.cs:
```diff
@@ -0,0 +1,86 @@
+    public class ManiaModNoLongNotes : Mod, IApplicableAfterBeatmapConversion
+    {
+
+        public override string Name => "No Long Notes";
```
Name feels a bit too long, maybe "Taps Only" or something?

bdach replied:
not sure "taps" fits with the rest of the mania vocabulary. the notes are technically called "hold" notes most everywhere else so that is probably preferred. you could even call this the "No Holds" mod.

Nivalyx replied:
fwiw in osu!mania the term for this is "rice" so if any i'd like to suggest `Rice Only` or sth like that

![image](https://user-images.githubusercontent.com/36564236/151410440-316371d0-0a2a-4229-ba86-2a261abfd0e2.png)

bdach replied:
"rice" is a colloquial / community term, and i personally found it was very confusing when i first encountered it. i don't see any references to that term [on the wiki](https://osu.ppy.sh/home/search?mode=wiki_page&query=rice) either, outside of tournament articles.

Walavouchey replied:
seems like we'll need to write stubs on these and give the community terms a mention there

![image](https://user-images.githubusercontent.com/36758269/151411371-3d527d03-e241-469a-9e15-d7d7cdb14d6f.png)

Spooghetti420 replied:
Rice Only does have a nice ring to it. 

> "rice" is a colloquial / community term, and i personally found it was very confusing when i first encountered it. 
With a proper description, it might be possible to avoid any confusion, if the mod clearly states what it means. That said, maybe the name is still a bit too colloquial? No Holds is a solid improvement. I would change it to that if people agree

peppy replied:
I’m okay with witty names but let’s not use “rice”.

frenzibyte commented on osu.Game.Rulesets.Mania/Mods/ManiaModNoLongNotes.cs:
```diff
@@ -0,0 +1,86 @@
+            var maniaBeatmap = (ManiaBeatmap)beatmap;
+
+            var newObjects = new List<ManiaHitObject>();
+            beatmap.HitObjects.OfType<HoldNote>().ForEach(h => 
```
nit: `ForEach` is designed usually for one-liner operations, but in this case, a regular `foreach` pattern should be used instead.

frenzibyte reviewed (changes requested):
Thanks for the contribution.

I have made a few notes on the code, in addition to CI results. It would be best to have basic test coverage for this mod (see all `osu.Game.Rulesets.Mania.Tests/Mods/TestSceneManiaMod*.cs` for example)

bdach commented on osu.Game.Rulesets.Mania/Mods/ManiaModNoLongNotes.cs:
```diff
@@ -0,0 +1,86 @@
+        
+        public override double ScoreMultiplier => 1;
+        
+        public override string Description => @"Turns all held notes into tap notes. No coordination required.";
```
as above, "tap" notes aren't really a term that gets used i don't think

```suggestion
        public override string Description => @"Turns all held notes into normal notes. No coordination required.";
```

Spooghetti420 replied:
How might I run inspectcode?
Also, under CONTRIBUTING.md it warns of pushing too many times to avoid burdening the CI queue. I'd gladly push your suggestion here if you'd confirm it's not a problem (otherwise I will implement all of the suggestions in this discussion and push them at once).

bdach replied:
for running inspectcode, see [readme](https://github.com/ppy/osu#code-analysis). if you're on a non-windows platform `InspectCode.sh` is also provided in the repository root.

as for the suggestion - it is fine to apply it locally as well and push out all required changes at once, not necessary to commit it via github.

bdach commented on osu.Game.Rulesets.Mania/Mods/ManiaModNoLongNotes.cs:
```diff
@@ -0,0 +1,86 @@
+                {
+                    Column = h.Column,
+                    StartTime = h.StartTime,
+                    Samples = h.Samples
```
i believe that these should be using `NodeSamples` rather than `Samples`. see how the head and tail get created here:

https://github.com/ppy/osu/blob/6beb0ff2a414899ee7be73eb7fd534f297933cd1/osu.Game.Rulesets.Mania/Objects/HoldNote.cs#L101-L113

bdach reviewed (commented):
there will be code style inspections due to too many empty lines here - please run inspectcode locally to resolve

Spooghetti420 committed:
Update mod per discussion + create test

Spooghetti420 committed:
Fix code formatting

Spooghetti420 commented:
Thanks for the suggestions, everyone. I've implemented all of the improvements for now, except the test "suite" is still a little lacking :-) It took a good deal to get it running, so maybe I'll return to it to add some more tests tomorrow, though I still don't know what test ideas to try out.
On a separate note — apologies for once again screwing up the formatting. I ran `dotnet format` before committing, but I guess that doesn't correct every single issue. I could push a commit to fix this, but I'm unsure as to whether this is a good idea.

smoogipoo commented on osu.Game.Rulesets.Mania/Mods/ManiaModNoHolds.cs:
```diff
@@ -0,0 +1,75 @@
+
+namespace osu.Game.Rulesets.Mania.Mods
+{
+    public class ManiaModNoHolds : Mod, IApplicableAfterBeatmapConversion
```
This one's the old version of this mod right? This file can be removed. Also, `ManiaRuleset` doesn't use the other mod.

Spooghetti420 replied:
No, this is the new one, since the old "No Long Notes" name was too long. I did delete the `ManiaModNoLongNotes.cs`, locally before committing, though, so I don't understand why it still appears after pushing.

peppy replied:
Following discussion in #difficulty-mania, let's go with the name "Hold Off".

Spooghetti420 replied:
This is a wonderful name!

smoogipoo commented on osu.Game.Rulesets.Mania.Tests/Mods/TestSceneManiaModNoHolds.cs:
```diff
@@ -0,0 +1,52 @@
+
+namespace osu.Game.Rulesets.Mania.Tests.Mods
+{
+    public class TestSceneManiaModNoHolds : ModTestScene
```
Likewise, this should be updated to match the new mod's name.

PercyDan54 commented on osu.Game.Rulesets.Mania.Tests/Mods/TestSceneManiaModNoHolds.cs:
```diff
@@ -0,0 +1,52 @@
+        public void TestMapHasNoHeldNotes()
+        {
+            var testBeatmap = createBeatmap();
+            Assert.That(!testBeatmap.HitObjects.OfType<HoldNote>().Any());
```
```suggestion
            Assert.False(testBeatmap.HitObjects.OfType<HoldNote>().Any());
```

Spooghetti420 replied:
Thanks, didn't know that was a thing

Spooghetti420 committed:
Merge branch 'master' of https://github.com/ppy/osu into no-long-notes-mod

Spooghetti420 committed:
Change threshold from ms to beat-based, add tests

Spooghetti420 commented:
Just changed the "threshold" we had before to derive from the beat value of a note rather than from its time in ms, so now only notes that last over 1/2 a beat (by default) get an end note. (Is the way I implemented it a good way to get a hold note's note value, though?)

Like @peppy and @smoogipoo said, it might be too excessive to have this kind of control specifically over end notes, but I included them anyway to see whether this setting is any good. Currently there's a dropdown that allows 1/1, 1/2, 1/4, etc. to be set as the thresholds, though please scrutinize this feature if it's not really appropriate.
Likewise, I've written several more tests with this setting concept in mind, though I cannot see them being useful if we remove the dropdown. I cannot guarantee the code is up to standard, but I've done my best for now. Thanks to everyone for all the feedback once again :)

peppy committed:
Apply code style changes

peppy committed:
Rename function and make `public` again for test usage

peppy committed:
Fix inspections in test scene

Spooghetti420 committed:
Remove settings

peppy committed:
Fix constant formatting

smoogipoo commented:
I think this mod should be made incompatible with `ManiaModInvert`, and be grouped under the same button.

peppy commented:
Agree with incompatible. Disagree with same button (they are functionally very different and should be top-level visible).

Spooghetti420 committed:
Make incompatible with

Spooghetti420 commented:
I also don't know about grouping under the same button, but I have gone and made incompatible with Invert. I also did wonder whether they should be incompatible at the beginning, but I wasn't sure enough to commit it.

bdach commented on osu.Game.Rulesets.Mania/Mods/ManiaModHoldOff.cs:
```diff
@@ -0,0 +1,72 @@
+
+        public static double GetNoteDurationInBeatLength(HoldNote holdNote, ManiaBeatmap beatmap)
+        {
+            double bpmAtNoteTime = beatmap.ControlPointInfo.TimingPointAt(holdNote.StartTime).BPM;
+            return (60 * holdNote.Duration) / (1000 * bpmAtNoteTime);
```
while this appears to somehow work, it is really complicated for some reason, to the point where i have no idea what the constants are doing - surely it'd be easier to write

```suggestion
            double beatLength = beatmap.ControlPointInfo.TimingPointAt(holdNote.StartTime).BeatLength;
            return holdNote.Duration / beatLength;
```

or is there some subtlety i'm missing? my version appears to pass tests, at least.

Spooghetti420 replied:
Looks like a big logical improvement, lol. The original logic was based on the thought that at 60BPM, 1 beat = 1 second, at 120 BPM 1 beat = 0.5 s, etc., so we had (duration*60) / (BPM*1000) = value in beats. As you said, even formulating an explanation for this is rather difficult; I simply arranged the terms until the formula worked, to be frank. It didn't occur to me to use the beatlength, but indeed this ought to be the same. Thanks for the improvement.

Spooghetti420 committed:
Improve beat length logic

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

peppy committed:
Update description

smoogipoo commented:
For future time travelers who may be interested in improving/adjusting this, my intentions regarding beat length weren't implemented as I suggested them to be. Rather than adding an ending note when the length exceeds some beat length value, my intention was to add a note _every X_ beat length.

It basically results in the following difference:
Original map: https://user-images.githubusercontent.com/1329837/152097084-a4b3dda6-d9c6-440f-80dd-ae85e2d69eff.mp4
This PR: https://user-images.githubusercontent.com/1329837/152097070-9aea110e-480d-47f9-9495-c0a24de8ae85.mp4
This PR with `END_NOTE_ALLOW_THRESHOLD = 1`: https://user-images.githubusercontent.com/1329837/152097048-f51bf7bc-254b-4219-91ef-258b8b357ed5.mp4
My suggestion with adding notes every 1 beatlength: https://user-images.githubusercontent.com/1329837/152097105-b8280067-bfe6-4666-8e68-2f3fdcaa2bfa.mp4

This takes advantage of two facts:
- Typically holds (whether in o!m or in osu!) end slightly before the next note, either 1/4 or 1/2 beat away, which allows the next note to arrive on the 1/1 beat.
- Holds in mania are typically hitsound-less for this same reason.

peppy commented:
Sounds better. Could notes not just be added at every tick?

smoogipoo commented:
Haven't explored that/don't know if mania properly uses tick rate at all (because it's not a thing in osu-stable), but possibly.
